File: testdata/precision-suite-ruby/tp_pathtraversal.rb
# Path traversal: a request-controlled filename flows into File.read with no
# containment check, so "../../etc/passwd" escapes the intended directory —
# CWE-22. A correct scanner fires TAINT-004. The File.basename-sanitized form in
# clean_safe_db.rb must NOT fire.
class DownloadsController
  def show
    name = params[:file]
    data = File.read("/srv/downloads/#{name}") # nox-expect: TAINT-004
    send_data data
  end

  def open_config
    path = params[:path]
    contents = File.open(path).read # nox-expect: TAINT-004
    render plain: contents
  end
end

File: testdata/precision-suite-php/tp_pathtraversal.php
<?php
// Path traversal / local file disclosure: a request parameter is passed to
// readfile() unchecked, so `?file=../../etc/passwd` escapes the intended
// directory. CWE-22, TAINT-004.

function download()
{
    $file = $_GET['file'];
    readfile("/var/www/files/" . $file); // nox-expect: TAINT-004
}
